File: ocaml/triangle.ml
(* Triangle utilities. *)

open Batteries

type box = { p1 : int; p2 : int; q1 : int; q2 : int }

let make_box p1 p2 q1 q2 = { p1; p2; q1; q2 }

let initial_box = make_box 1 1 2 3

let children box =
  let x = box.p2 in
  let y = box.q2 in
  [ make_box (y-x) x y     (y*2 - x);
    make_box x     y (x+y) (x*2 + y);
    make_box y     x (y+x) (y*2 + x) ]

type triple = { a : int; b : int; c : int }
type t = triple

let compare = compare

let circumference triangle = triangle.a + triangle.b + triangle.c

(* Return the three sides of the pythagorean triple described by the
   given box. *)
let box_triangle box =
  { a = box.q1 * box.p1 * 2;
    b = box.q2 * box.p2;
    c = box.p1 * box.q2 + box.p2 * box.q1 }

let multiply_triple tri k =
  { a = tri.a * k; b = tri.b * k; c = tri.c * k }

(* Generate all of the primitive Pythagorean triples with a
   circumference <= limit.  Calls 'act triple circumference' for each
   possible triple. *)
let generate_fibonacci_triples limit act =
  let rec loop work = match work with
      [] -> ()
    | (box :: nextWork) ->
      let triple = box_triangle box in
      let size = circumference triple in
      if size <= limit then begin
	act triple size;
	loop (nextWork @ children box)
      end in
  loop [initial_box]

let generate_triples limit act =
  let sub_generate triple _size =
    let rec loop k =
      let k_triple = multiply_triple triple k in
      let k_size = circumference k_triple in
      if k_size <= limit then begin
	act k_triple k_size;
	loop (k + 1)
      end in
    loop 1 in
  generate_fibonacci_triples limit sub_generate

File: go/src/euler/triangle.go
package euler

// Triangle manipulations.
type Box struct {
	P1, P2, Q1, Q2 int
}

type Triple struct {
	A, B, C int
}

type TripleInfo struct {
	Tri Triple
	Circumference int
}

var initialBox = Box{P1: 1, P2: 1, Q1: 2, Q2: 3}

// Compute the primitive triples.
func FibonacciTriples(limit int) (result []TripleInfo) {
	result = make([]TripleInfo, 0)

	work := make([]Box, 1)
	work[0] = initialBox

	for len(work) > 0 {
		box := work[0]
		work = work[1:]

		tri := makeTriple(box)
		circ := ComputeCircumference(tri)
		if circ <= limit {
			work = append(work, children(box)...)
			result = append(result, TripleInfo{Tri: tri, Circumference: circ})
		}
	}
	return
}

func AllTriples(limit int) (result []TripleInfo) {
	base := FibonacciTriples(limit)
	result = make([]TripleInfo, 0)

	for _, t := range base {
		for k := 1; true; k++ {
			t2 := multipleTriple(t.Tri, k)
			c2 := t.Circumference * k
			if c2 > limit {
				break
			}

			result = append(result, TripleInfo{Tri: t2, Circumference: c2})
		}
	}
	return
}

func children(in Box) (result []Box) {
	result = make([]Box, 3)

	x := in.P2
	y := in.Q2

	result[0] = Box{P1: y - x, P2: x, Q1: y, Q2: y * 2 - x}
	result[1] = Box{P1: x, P2: y, Q1: x + y, Q2: x * 2 + y}
	result[2] = Box{P1: y, P2: x, Q1: x + y, Q2: y * 2 + x}
	return
}

func ComputeCircumference(tri Triple) int {
	return tri.A + tri.B + tri.C
}

func makeTriple(in Box) (result Triple) {
	result.A = in.Q1 * in.P1 * 2
	result.B = in.Q2 * in.P2
	result.C = in.P1 * in.Q2 + in.P2 * in.Q1
	return
}

func multipleTriple(in Triple, k int) (result Triple) {
	result.A = in.A * k
	result.B = in.B * k
	result.C = in.C * k
	return
}
